### PR TITLE
Reduce number of shadow calls

### PR DESCRIPTION
**Pull request by rollingtumbling**
Greatly reduced the number of shadow calls for
- Health check on unhealthy providers
- Latency check on non-selected healthy providers
Both type of shadow calls now strictly follow "at most 1 call every 15 seconds" rule.
This should help avoid latency increase when launch RPC gateway to production.

We can see the shadow call becomes way way less than actual request-serving call.
Screenshot 2024-03-22 at 2 58 32 AM

### CODE REVIEW

**Comment by rollingtumbling** [comment]
* **#525** <a href="https://app.graphite.dev/github/pr/Uniswap/routing-api/525?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rollingtumbling and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lib/rpc/SingleJsonRpcProvider.ts [hunk 1]
```diff
@@ -76,8 +89,6 @@ export class SingleJsonRpcProvider extends StaticJsonRpcProvider {
         this.config.LATENCY_STAT_HISTORY_WINDOW_LENGTH_IN_S,
         log
       )
-      // Fire and forget. Won't check the sync result.
-      this.maybeSyncAndUpdateProviderState()
```
Comment (rollingtumbling) [inline lib/rpc/SingleJsonRpcProvider.ts:80]: This is to reduce the time to initialize lambda.
The first sync is actually optional.

**Review by jsy1218 — APPROVED** [review]

File: bin/stacks/rpc-gateway-dashboard.ts [hunk 2]
```diff
@@ -51,26 +52,18 @@ function getHealthScoreMetricsForChain(chainId: ChainId) {
...
-      },
-    ])
+    for (const methodName of MAJOR_METHOD_NAMES) {
+      metrics.push([
+        'Uniswap',
+        `RPC_GATEWAY_${chainId}_${providerName}_evaluated_latency_${methodName}`,
+        'Service',
+        'RoutingAPI',
+        {
+          id: `getBlockNumber_latency_${chainId}_${providerName}`,
```
[the remark this answers is not in the document]
Comment (rollingtumbling) [reply bin/stacks/rpc-gateway-dashboard.ts]: You are right. Sorry about that! Fixed

File: lib/rpc/SingleJsonRpcProvider.ts [hunk 3]
```diff
@@ -89,106 +100,147 @@ export class SingleJsonRpcProvider extends StaticJsonRpcProvider {
...
     this.healthScore += this.config.ERROR_PENALTY
     this.log.error(
-      `${this.url}: method: ${method} error penalty ${this.config.ERROR_PENALTY}, score => ${this.healthScore}`
+      `${this.url}: method: ${perf.methodName} error penalty ${this.config.ERROR_PENALTY}, score => ${this.healthScore}`
     )
   }
 
-  private recordHighLatency(method: string) {
+  private recordHighLatency(perf: SingleCallPerf) {
+    if (perf.callType === CallType.HEALTH_CHECK) {
```
[the remark this answers is not in the document]
Comment (rollingtumbling) [reply lib/rpc/SingleJsonRpcProvider.ts:145]: The design is that high latency is a type of error that will penalize the provider. A response without high latency is however treated as a successful sanity check that will reward the provider. So the logic is a little different there for health score update. For latency records, they are the same.

**Review by cgkol — APPROVED** [review]
lets give it a go, this code is getting convoluted for sure but hopefully its fine. THe way we set true and false the latency and healthiness is kind of hard to keep track off but hopefully it all works out

**Pull request merged**